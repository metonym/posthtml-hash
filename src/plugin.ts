import fs from 'fs';
import path from 'path';
import { PostHTML } from 'posthtml';
import { createHash, hashFileName, processFile } from './utils';

const DEFAULT_PATH = '';
const DEFAULT_HASH_LENGTH = 20;
const DEFAULT_OPTIONS: IOptions = {
  path: DEFAULT_PATH,
  hashLength: DEFAULT_HASH_LENGTH
};

function plugin(options = DEFAULT_OPTIONS) {
  return function posthtmlHash(tree: PostHTML.Node) {
    const nonEmptyString = new RegExp(/\S+/);

    const matchers: IPostHTMLHashMatcher[] = [
      { tag: 'link', attrs: { rel: 'stylesheet', href: nonEmptyString } },
      { tag: 'script', attrs: { src: nonEmptyString } }
    ];

    tree.match(matchers, node => {
      const attrs = node.attrs!;
      let fileName = '';

      if (attrs.href) {
        fileName = attrs.href;
      } else if (attrs.src) {
        fileName = attrs.src;
      }

      const pathToFile = options.path || '';
      const file = path.join(process.cwd(), pathToFile, fileName);

      processFile(file, () => {
        const buffer = fs.readFileSync(file);
        const hashLength = options.hashLength || DEFAULT_HASH_LENGTH;
        const hash = createHash(buffer, hashLength);
        const hashedFileName = hashFileName(fileName, hash);
        const hashedFile = path.join(process.cwd(), pathToFile, hashedFileName);

        fs.renameSync(file, hashedFile);

        if (attrs.href) {
          node.attrs!.href = hashedFileName;
        } else if (attrs.src) {
          node.attrs!.src = hashedFileName;
        }
      });

      return node;
    });
  };
}

interface IOptions {
  path?: string;
  hashLength?: number;
}

interface IPostHTMLHashMatcher {
  tag: PostHTML.StringMatcher;
  attrs: PostHTML.AttrMatcher;
}

export { plugin, DEFAULT_HASH_LENGTH };
