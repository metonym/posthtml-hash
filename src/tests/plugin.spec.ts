import posthtml from 'posthtml';
import plugin from '../';

describe('posthtml-hash', () => {
  it('matches the snapshot – default options', () => {
    posthtml()
      .use(plugin({ path: 'src/tests/__fixtures__' }))
      .process(
        `
        <html>
          <head>
            <link rel="stylesheet" href="bundle.min.css" />
          </head>
          <body>
            <script src="bundle.min.js"></script>
          </body>
        </html>
      `
      )
      .then((result: { html: string }) => {
        expect(result.html).toMatchSnapshot();
      });
  });

  it('matches the snapshot – custom hash length', () => {
    posthtml()
      .use(plugin({ path: 'src/tests/__fixtures__', hashLength: 10 }))
      .process(
        `
        <html>
          <head>
            <link rel="stylesheet" href="bundle.min.css" />
          </head>
          <body>
            <script src="bundle.min.js"></script>
          </body>
        </html>
      `
      )
      .then((result: { html: string }) => {
        expect(result.html).toMatchSnapshot();
      });
  });
});
