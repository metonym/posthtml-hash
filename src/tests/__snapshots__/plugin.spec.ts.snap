// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`posthtml-hash matches the snapshot – custom hash length 1`] = `
"
        <html>
          <head>
            <link rel=\\"stylesheet\\" href=\\"bundle.min.9a6cf95c41.css\\">
          </head>
          <body>
            <script src=\\"bundle.min.b0dcc67ffc.js\\"></script>
          </body>
        </html>
      "
`;

exports[`posthtml-hash matches the snapshot – default options 1`] = `
"
        <html>
          <head>
            <link rel=\\"stylesheet\\" href=\\"bundle.min.9a6cf95c41e87b9dc102.css\\">
          </head>
          <body>
            <script src=\\"bundle.min.b0dcc67ffc1fd562f212.js\\"></script>
          </body>
        </html>
      "
`;
